### PR TITLE
[NVPTX] Add TMA bulk tensor reduction intrinsics

### PR DESCRIPTION
**Pull request by durga4github**
This patch adds NVVM intrinsics and NVPTX codegen for:
* cp.async.bulk.tensor.reduce.1D -> 5D variants, supporting both Tile and Im2Col modes.
* These intrinsics optionally support cache_hints as indicated by the boolean flag argument.
* Lit tests are added for all combinations of these intrinsics in cp-async-bulk-tensor-reduce.ll.
* The generated PTX is verified with a 12.3 ptxas executable.
* Added docs for these intrinsics in NVPTXUsage.rst file.

PTX Spec reference:
https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cp-reduce-async-bulk-tensor

### CODE REVIEW

**Comment by durga4github** [comment]
Ping,
@Artem-B , Could you please help with this review?

File: llvm/docs/NVPTXUsage.rst [hunk 1]
```diff
@@ -663,6 +663,84 @@ the same functionality as described in the ``tile`` mode intrinsics above.
...
+  ========== =============
+  ``0``       ADD
+  ``1``       MIN
+  ``2``       MAX
+  ``3``       INC
+  ``4``       DEC
+  ``5``       AND
+  ``6``       OR
+  ``7``       XOR
+  ========== =============
```
[the remark this answers is not in the document]
Comment (durga4github) [reply llvm/docs/NVPTXUsage.rst]: Sure, I made the reduction_op as part of the intrinsic name.
Please let me know if the latest revision looks good.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 2]
```diff
@@ -4157,9 +4157,9 @@ bool NVPTXScopes::empty() const { return Scopes.size() == 0; }
        ? NVPTX::CP_ASYNC_BULK_TENSOR_##dir##_##dim##_SHARED32_##mode##suffix   \
        : NVPTX::CP_ASYNC_BULK_TENSOR_##dir##_##dim##_##mode##suffix)
 
-#define GET_CP_ASYNC_BULK_TENSOR_OPCODE_S2G(dim, mode)                         \
-  (IsCacheHint ? (CP_ASYNC_BULK_TENSOR_OPCODE(S2G, dim, mode, _CH))            \
-               : (CP_ASYNC_BULK_TENSOR_OPCODE(S2G, dim, mode, )))
+#define GET_CP_ASYNC_BULK_TENSOR_OPCODE_CH(op, dim, mode)                      \
+  (IsCacheHint ? (CP_ASYNC_BULK_TENSOR_OPCODE(op, dim, mode, _CH))             \
```
[the remark this answers is not in the document]
Comment (durga4github) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Yes, I made all the macros self-contained in the latest revision.

Resolving.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 3]
```diff
@@ -4177,31 +4177,40 @@ bool NVPTXScopes::empty() const { return Scopes.size() == 0; }
...
 static unsigned GetCpAsyncBulkTensorS2GOpcode(size_t Dim, bool IsShared32,
-                                              bool IsCacheHint, bool IsIm2Col) {
+                                              bool IsCacheHint, bool IsIm2Col,
+                                              bool IsReduce = false) {
   if (IsIm2Col) {
     switch (Dim) {
     case 3:
-      return GET_CP_ASYNC_BULK_TENSOR_OPCODE_S2G(3D, IM2COL);
+      return IsReduce ? GET_CP_ASYNC_BULK_TENSOR_OPCODE_CH(RED, 3D, IM2COL)
+                      : GET_CP_ASYNC_BULK_TENSOR_OPCODE_CH(S2G, 3D, IM2COL);
```
[the remark this answers is not in the document]
Comment (durga4github) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Handled in the latest revision,
Resolving this,

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td [hunk 4]
```diff
@@ -596,11 +598,37 @@ multiclass CP_ASYNC_BULK_TENSOR_S2G_INTR<int dim, bit shared32, string mode> {
...
+multiclass CP_ASYNC_BULK_TENSOR_REDUCE_INTR<int dim, bit shared32, string mode> {
+  defvar dims_dag = !dag(ins, !listsplat(Int32Regs, dim), !foreach(i, !range(dim), "d" # i));
+  defvar dims_str = !interleave(!foreach(i, !range(dim), "$d" # i), ", ");
+  defvar asm_str = " [$tmap, {{" # dims_str # "}}], [$src]";
+  defvar rc = !if(shared32, Int32Regs, Int64Regs);
+
+  defvar prefix = "cp.reduce.async.bulk.tensor" # "." # dim # "d" # ".global.shared::cta";
+  defvar suffix = "." # mode # ".bulk_group";
+
+  def "": NVPTXInst<(outs),
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:615]: I think you can use the magic `NAME` name here which may look a bit less odd than `""`.

```
def NAME: ...
def NAME # _CH: ...
```

**Review by Artem-B — APPROVED** [review]
LGTM with a minor nit.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td (hunk 4 above)
Comment (durga4github) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td]: Sure, I fixed it in the latest revision, only for the set of intrinsics that this PR adds.
I will send a separate NFC PR to do a similar change for other TMA ones.

And, Thanks for approval, Artem!

I will merge once the build results are clean.

**Comment by durga4github** [comment]
@kazutakahirata 
I was writing a fix ;-)

Thanks for merging the fix.
However, instead of removing the unreachable altogether, we can have it outside the switch-case.
Would you amend ? or shall I submit a separate PR for this?

**Comment by kazutakahirata** [comment]
> Thanks for merging the fix. However, instead of removing the unreachable altogether, we can have it outside the switch-case. Would you amend ? or shall I submit a separate PR for this?

If you could send me a separate PR, I'll LGTM it right away.  Thanks!

**Comment by durga4github** [comment]
> > Thanks for merging the fix. However, instead of removing the unreachable altogether, we can have it outside the switch-case. Would you amend ? or shall I submit a separate PR for this?
> 
> If you could send me a separate PR, I'll LGTM it right away. Thanks!

https://github.com/llvm/llvm-project/pull/117850

**Pull request merged**